csharp/src/IceDiscovery/Lookup.cs

```diff
@@ -162,354 +75,240 @@ public void FindAdapterById(string domainId, string adapterId, ILookupReplyPrx?
+                    task = InvokeFindAdapterAsync(id, replyServant);
                 }
-
-                var task = new TaskCompletionSource<IObjectPrx?>();
-                if (request.AddCallback(task))
+                else
                 {
-                    try
-                    {
-                        request.Invoke(_domainId, _lookups);
-                        Timer.Schedule(request, _timeout);
-                    }
-                    catch
-                    {
-                        request.Finished(null);
-                        _adapterRequests.Remove(adapterId);
-                    }
+                    task = replyServant.CompletionSource.Task;
                 }
-                return new ValueTask<IObjectPrx?>(task.Task);
             }
+            return await task.ConfigureAwait(false);
         }
 
-        internal ValueTask<IObjectPrx?> FindObjectAsync(Identity id)
+        internal async Task<IObjectPrx?> InvokeFindAdapterAsync(string id, LookupReply replyServant)
         {
-            lock (_mutex)
-            {
-                if (!_objectRequests.TryGetValue(id, out ObjectRequest? request))
-                {
-                    request = new ObjectRequest(this, id, _retryCount);
-                    _objectRequests.Add(id, request);
-                }
+            // Called with mutex locked
+            _adapterReplies.Add(id, replyServant);
+            Identity requestId = _replyAdapter.AddWithUUID(replyServant, ILocatorRegistryPrx.Factory).Identity;
 
-                var task = new TaskCompletionSource<IObjectPrx?>();
-                if (request.AddCallback(task))
+            Task<IObjectPrx?> replyTask = replyServant.CompletionSource.Task;
+            try
+            {
+                for (int i = 0; i < _retryCount; ++i)
                 {
-                    try
+                    long start = DateTime.Now.Ticks;
+                    int failureCount = 0;
+                    foreach ((ILookupPrx lookup, ILookupReplyPrx? reply) in _lookups)
                     {
-                        request.Invoke(_domainId, _lookups);
-                        Timer.Schedule(request, _timeout);
+                        ILookupReplyPrx? lookupReply = reply.Clone(requestId, ILookupReplyPrx.Factory);
+                        try
+                        {
+                            await lookup.FindAdapterByIdAsync(_domainId, id, lookupReply).ConfigureAwait(false);
+                        }
+                        catch (Exception ex)
+                        {
+                            if (++failureCount == _lookups.Count)
+                            {
+                                _lookup.Communicator.Logger.Warning(
+                                    $"failed to lookup adapter `{id}' with lookup proxy `{_lookup}':\n{ex}");
+                                replyServant.CompletionSource.SetResult(null);
+                            }
+                        }
```

Comment:
It's no longer using _warnOnce? This loop could be used by both findObject/findAdapter if moved to a private method that takes a lambda.

Reply:
I updated this to use a single method and a lambda, I removed `_warnOnce` my concern is that if you miss the first warning you don't know why replies are missing, even if the exceptions are always similar seems safer to always log the warning.